scheds/rust/scx_rustlite/src/main.rs

```diff
@@ -0,0 +1,314 @@
+    }
+
+    fn pop(&mut self) -> Option<Task> {
+        self.tasks.pop_first()
+    }
+}
+
+// Main scheduler object
+struct Scheduler<'a> {
+    skel: BpfSkel<'a>,
+    task_pool: TaskPool,
+    struct_ops: Option<libbpf_rs::Link>,
+}
+
+impl<'a> Scheduler<'a> {
+    fn init(opts: &Opts) -> Result<Self> {
+        // Open the BPF prog first for verification.
+        let skel_builder = BpfSkelBuilder::default();
+        let mut skel = skel_builder.open().context("Failed to open BPF program")?;
+        let pid = std::process::id();
+
+        // Scheduler task pool to sort task by vtime.
+        let task_pool = TaskPool::new();
+
+        // Set scheduler options (defined in the BPF part)
+        skel.bss_mut().usersched_pid = pid;
+        skel.rodata_mut().slice_ns = opts.slice_us * 1000;
+        skel.rodata_mut().switch_partial = opts.partial;
+
+        // Attach BPF scheduler.
+        let mut skel = skel.load().context("Failed to load BPF program")?;
+        skel.attach().context("Failed to attach BPF program")?;
+        let struct_ops = Some(
+            skel.maps_mut()
+                .rustlite()
+                .attach_struct_ops()
+                .context("Failed to attach struct ops")?,
+        );
+        info!("{} scheduler attached", SCHEDULER_NAME);
+
+        // Return scheduler object.
+        Ok(Self {
+            skel,
+            task_pool,
+            struct_ops,
+        })
+    }
+
+    fn read_bpf_exit_kind(&mut self) -> i32 {
+        unsafe { std::ptr::read_volatile(&self.skel.bss().exit_kind as *const _) }
+    }
+
+    fn report_bpf_exit_kind(&mut self) -> Result<()> {
+        match self.read_bpf_exit_kind() {
+            0 => Ok(()),
+            etype => {
+                let cstr = unsafe { CStr::from_ptr(self.skel.bss().exit_msg.as_ptr() as *const _) };
+                let msg = cstr
+                    .to_str()
+                    .context("Failed to convert exit msg to string")
+                    .unwrap();
+                info!("BPF exit_kind={} msg={}", etype, msg);
+                Ok(())
+            }
+        }
+    }
+
+    fn schedule(&mut self) {
+        let maps = self.skel.maps();
+        let enqueued = maps.enqueued();
+        let dispatched = maps.dispatched();
+
+        // Drain enqueued list and push all the tasks to the task pool (sorted by vtime).
+        loop {
+            match enqueued.lookup_and_delete(&[]) {
+                Ok(Some(val)) => {
+                    let task = unsafe { &*(val.as_slice().as_ptr() as *const bpf_intf::task_ctx) };
+                    self.task_pool.push(task.pid, task.vtime);
+                }
+                Ok(None) => break,
+                Err(err) => {
+                    warn!("Error: {}", err);
+                    break;
+                }
+            }
+        }
+
+        // Dispatch tasks from the task pool in order (sending them to the BPF part).
+        loop {
+            match self.task_pool.pop() {
```

Comment:
I know the focus is on being easy but I wonder whether the fact that all tasks are immediately dispatched is too limiting. We want dispatch to be happening when the CPUs are ready to run tasks so that tasks can be queued and ordered while waiting for CPUs to open up. Maybe the interlocking is a bit tricky to figure out but it might not take a lot of code for the BPF part to indicate how many CPUs are currently looking for tasks and userspace part can dispatch accordingly. It doesn't have to be completely accurate either.

Reply:
Maybe if I move the whole `select_cpu()` logic to user-space this part should also become easier... need to think a little bit more about this.